include/networkit/graph/Graph.hpp

```diff
@@ -139,6 +145,389 @@ class Graph final {
+    public:
+        NodeAttributeStorage(const Graph *theGraph, std::string name)
+            : NodeAttributeStorageBase{theGraph, std::move(name), typeid(T)} {}
+
+        ~NodeAttributeStorage() override { invalidateAttributes(); }
+
+        void invalidateAttributes() override {
+            for (auto att : attrSet)
+                att->invalidateAttribute();
+        }
+
+        void resize(index i) {
+            if (i >= values.size()) {
+                values.resize(i + 1);
+            }
+        }
+
+        auto size() { return validElements; }
+
+        void set(index i, T v) {
+            markValid(i);
+            resize(i);
+            values[i] = std::move(v);
+        }
+
+        // instead of returning an std::optional (C++17) we provide these
+        // C++14 options
+        // (1) throw an exception when invalid:
+        T get(index i) { // may throw
+            checkIndex(i);
+            return values[i];
+        }
+
+        // (2) give default value when invalid:
+        T get(index i, T defaultT) {
+            if (i >= values.size() || !isValid(i)) {
+                return defaultT;
+            }
+            return values[i];
+        }
+
+    private:
+        using NodeAttributeStorageBase::theGraph;
+        std::vector<T> values;
+        friend class NodeAttribute<T>;
+        std::unordered_set<NodeAttribute<T> *> attrSet;
+    }; // class NodeAttributeStorage<T>
+
+    template <typename T>
+    class NodeAttribute {
+    public:
+        class Iterator {
+        public:
+            // The value type of the nodes (i.e. nodes). Returned by
+            // operator*().
+            using value_type = T;
+
+            // Reference to the value_type, required by STL.
+            using reference = value_type &;
+
+            // Pointer to the value_type, required by STL.
+            using pointer = value_type *;
+
+            // STL iterator category.
+            using iterator_category = std::forward_iterator_tag;
+
+            // Signed integer type of the result of subtracting two pointers,
+            // required by STL.
+            using difference_type = ptrdiff_t;
+
+            Iterator() : storage{nullptr}, idx{0} {}
+            Iterator(NodeAttributeStorage<T> *storage) : storage{storage}, idx{0} {
+                if (storage) {
+                    nextValid();
+                }
+            }
+
+            Iterator &nextValid() {
+                while (storage && !storage->isValid(idx)) {
+                    if (idx >= storage->values.size()) {
+                        storage = nullptr;
+                        return *this;
+                    }
+                    ++idx;
+                }
+                return *this;
+            }
+
+            Iterator &operator++() {
+                if (!storage) {
+                    throw std::runtime_error("Invalid attribute iterator");
+                }
```

Comment:
As above, we should expect the user to use the iterator correctly. I recommend replacing this with an assertion.

Reply:
See above (https://github.com/networkit/networkit/discussions/769).